gridfs/grid_file.py

```diff
@@ -504,30 +504,26 @@ def read(self, size=-1):
         """
         self._ensure_file()
 
-        if size == 0:
-            return EMPTY
-
         remainder = int(self.length) - self.__position
         if size < 0 or size > remainder:
             size = remainder
 
+        if size == 0:
+            return EMPTY
+
         received = 0
         data = StringIO()
         while received < size:
             chunk_data = self.readchunk()
             received += len(chunk_data)
             data.write(chunk_data)
 
-        # Detect extra chunks.
-        max_chunk_n = math.ceil(self.length / float(self.chunk_size))
-        chunk = self.__chunks.find_one({"files_id": self._id,
-                                        "n": {"$gte": max_chunk_n}},
-                                       session=self._session)
-        # According to spec, ignore extra chunks if they are empty.
-        if chunk is not None and len(chunk['data']):
-            raise CorruptGridFile(
-                "Extra chunk found: expected %i chunks but found "
-                "chunk with n=%i" % (max_chunk_n, chunk['n']))
+        # Detect extra chunks after reading the entire file.
+        if size == remainder and self.__chunk_iter:
+            try:
+                self.__chunk_iter.next()
+            except StopIteration:
+                pass
```

Comment:
The gridfs spec has this to say:

> Should drivers report an error if a stored file has extra chunks?
>
> The length and the chunkSize fields of the files collection document together imply exactly how many chunks a stored file should have. If the chunks collection has any extra chunks the stored file is in an inconsistent state. Ideally we would like to report that as an error, but this is an extremely unlikely state and we don't want to pay a performance penalty checking for an error that is almost never there. Therefore, drivers MAY ignore extra chunks.

Given that, I've removed the check on every call to `read()` but left the check for after reading the entire file. Calling `next()` one extra time will have no performance impact in the common case (because the cursor is already exhausted) and still keeps the `CorruptGridFile` check.

Note that there is a minor difference here: An app used to see `CorruptGridFile("Extra chunk found..")` on the *first* `read()` of a file with extra chunks. Now the app will only see the error on the *final* `read()`. I think this is a totally acceptable change and is so minor I don't think it's necessary to call out in the docs.

Reply:
You should call it out in the changelog.

Reply:
I added some `versionchanged` entries and updated the changelog. I also tweaked the chunk iterator to retry once on CursorNotFound for backwards compatibility -- previously an app could spent an arbitrary amount of time between two `read()` calls because each used a `find_one`. However when using a cursor the server side cursor can timeout and thus the same code would not work. I will add a test for this on Monday.

Reply:
I added a new test that simulates a cursor timeout by killing the underlying cursor before a getMore call.